### PR TITLE
Add device to state

### PR DESCRIPTION
**Pull request by mvpatel2000**
# What does this PR do?

As title. Will help algorithms/callbacks move data around

# What issue(s) does this change relate to?

[CO-1448](https://mosaicml.atlassian.net/browse/CO-1448)

### CODE REVIEW

**Comment by mvpatel2000** [comment]
@alextrott16 tagging since it touches seq length warmup. Feel free to suggestion any other simplifications that arise from this.

File: composer/core/state.py [hunk 1]
```diff
@@ -139,6 +140,7 @@ class State(Serializable):
         algorithms (Algorithm | Sequence[Algorithm], optional): The algorithms used for training.
         callbacks (Callback | Sequence[Callback], optional): The callbacks used for training.
         deepspeed_config (Dict[str, Any], optional): The configuration dictionary for deepspeed.
+        device (Device): The device used by this process.
```
Comment (dakinggg) [inline composer/core/state.py:143]: I'm not sure whether the name should be changed, or just the docs more clear, but I think we need to be a bit more clear about what this device is. Technically you can use whatever device you want in your model code. I guess this is the device that we are going to move your model to, for this process?

**Review by dakinggg — APPROVED** [review]
LGTM

**Review by eracah — APPROVED** [review]
LGTM.

**Review by nik-mosaic — APPROVED** [review]

**Pull request merged**